tools/collect_hana_sr_diag.sh

```diff
@@ -0,0 +1,645 @@
+# Declare associative array for remote tool cache
+declare -gA REMOTE_TOOL_CACHE
+
+check_and_run_remote_command() {
+    local node_ip="$1"
+    local tool_name="$2"
+    local remote_command="$3"
+    local output_file="$4" # Optional: where to save stdout
+    local log_errors_to="$5" # Optional: where to save stderr
+    local cache_key="${node_ip}_${tool_name}"
+    local tool_check_result
+
+    # Special handling for hdbnsutil and HDBSettings: skip 'which' check if full path is provided in remote_command
+    if [[ "$tool_name" == "hdbnsutil" || "$tool_name" == "HDBSettings" ]]; then
+        tool_check_result=0
+        log_message "INFO" "Skipping 'which' check for '$tool_name' as full path is expected in remote command."
+        REMOTE_TOOL_CACHE[$cache_key]=$tool_check_result
+    else
+        # Determine if `which` needs sudo
+        local which_prefix=""
+        if [[ "$remote_command" == *sudo* ]]; then
+            which_prefix="sudo "
+        fi
+        local which_cmd="${which_prefix}which \\$tool_name"
+
+        # Only check each executable or tool once
+        if [[ -v REMOTE_TOOL_CACHE[$cache_key] ]]; then
+            tool_check_result=${REMOTE_TOOL_CACHE[$cache_key]}
+            #log_message "INFO" "Using cached result for '$tool_name' on $node_ip: $([[ $tool_check_result -eq 0 ]] && echo 'Found' || echo 'Not Found')."
+        else
+            log_message "INFO" "Checking if '$tool_name' exists on $node_ip (first check)..."
+            # shellcheck disable=SC2029
+            if ssh "${SSH_USER}@${node_ip}" "$which_cmd" &>/dev/null; then
+                tool_check_result=0
+                log_message "SUCCESS" "'$tool_name' found on $node_ip."
+            else
+                tool_check_result=1
+                log_message "ERROR" "'$tool_name' not found or not executable on $node_ip. Skipping command."
+            fi
+            REMOTE_TOOL_CACHE[$cache_key]=$tool_check_result
+        fi
+    fi
+    if [[ $tool_check_result -ne 0 ]]; then
+        # Tool not found, so we return immediately. No need to log "Skipping command" here,
+        # as it was logged when tool_check_result was determined.
+        return 1
+    fi
+
+    log_message "INFO" "Running: $remote_command on $node_ip"
+    local command_failed=0
+    if [[ -n "$output_file" ]]; then
+        # shellcheck disable=SC2029
+        if ! ssh "${SSH_USER}@${node_ip}" "$remote_command" > "$output_file" 2>> "${log_errors_to:-$LOG_FILE}"; then
+            command_failed=1
+        fi
+    else
+        # shellcheck disable=SC2029
+        if ! ssh "${SSH_USER}@${node_ip}" "$remote_command" 2>> "${log_errors_to:-$LOG_FILE}"; then
+            command_failed=1
+        fi
+    fi
+
+    if [[ $command_failed -ne 0 ]]; then
+        if [[ "$tool_name" == "HDBSettings" ]]; then
+            log_message "WARNING" "Remote command '$remote_command' on $node_ip returned non-zero exit code. Ignoring as requested for HDBSettings."
+            return 0
```

Comment:
I don't think this line is needed. In fact, perhaps the whole block can be re-written as:
```
        if [[ "$tool_name" != "HDBSettings" ]]; then
            log_message "ERROR" "Remote command '$remote_command' failed on $node_ip."
            return 1
        fi
        log_message "WARNING" "Remote command '$remote_command' on $node_ip returned non-zero exit code. Ignoring as requested for HDBSettings."
```